File: 1. Extension classes/src/airco_demo.cpp
#include "airco.h"

using namespace std::chrono_literals;

int main()
{
    std::cout << "Testing Airco class" << std::endl;

    auto callback = [](bool cooler_is_running) {
        std::cout << "[Callback] Cooler is" << (cooler_is_running ? "" : " not") << " running" << std::endl;
    };

    std::cout << "Creating airco without cooler and controller" << std::endl;
    auto airco = Airco();
    std::cout << "Trying to add callback function (should give error message)" << std::endl;
    airco.AddCallback(callback);
    std::cout << "Trying to start airco (should give error message)" << std::endl;
    airco.Start();
    std::cout << "Trying temperature values (should not give any result)" << std::endl;
    airco.Temperature(19.8);
    airco.Temperature(20.5);
    airco.Temperature(21.2);

    std::cout << "Creating controller" << std::endl;
    auto ctrl = std::make_shared<Controller>();
    std::cout << "Adding controller to airco" << std::endl;
    airco.SetController(ctrl);
    std::cout << "Trying to add callback function (should give error message)" << std::endl;
    airco.AddCallback(callback);
    std::cout << "Trying to start airco (should give error message)" << std::endl;
    airco.Start();
    std::cout << "Trying temperature values (should not give any result)" << std::endl;
    airco.Temperature(19.8);
    airco.Temperature(20.5);
    airco.Temperature(21.2);

    std::cout << "Creating cooler" << std::endl;
    auto cooler = std::make_shared<Cooler>();
    std::cout << "Adding cooler to airco" << std::endl;
    airco.SetCooler(cooler);
    std::cout << "Trying to add callback function" << std::endl;
    airco.AddCallback(callback);
    std::cout << "Trying to start airco (should be successful)" << std::endl;
    airco.Start();
    std::cout << "Trying temperature values" << std::endl;
    airco.Temperature(19.8);
    airco.Temperature(20.5);
    airco.Temperature(21.2);
    airco.Temperature(20.6);
    airco.Temperature(19.9);
    std::cout << "Stopping airco" << std::endl;
    airco.Stop();
    std::cout << "Trying temperature values (should not give any result)" << std::endl;
    airco.Temperature(19.8);
    airco.Temperature(20.5);
    airco.Temperature(21.2);

    return 0;
}

File: 1. Extension classes/src/airco_demo.py
from airco import Airco, Temperature, Thermostat, Controller


if __name__ == '__main__':
    print("Testing airco")

    print("Creating thermostat")
    thermostat = Thermostat()

    print("Creating controller")
    controller = Controller()

    print("Creating airco")
    airco = Airco(controller, thermostat)

    print("Adding callback functions")
    cb = lambda running: print(f"[Callback] airco is{'' if running else ' not'} cooling")
    airco.add_callback(cb)

    print("Starting airco")
    airco.start()
    airco.temperature(20.5)
    airco.temperature(21.2)
    airco.temperature(20.6)
    airco.temperature(19.9)
    airco.temperature(20.4)

    print("Stopping airco")
    airco.stop()

    del airco

    print("Testing inheritance")

    class PyController(Controller):
        def signal(self, event):
            print("PyController received event:", event)
            Controller.signal(self, event)


    py_controller = PyController()

    print("Creating airco with inherited controller")
    airco2 = Airco(py_controller, thermostat)
    airco2.add_callback(cb)
    airco2.start()
    print("Setting temperature")
    print("Expected: PyController received event: (but does not happen)")
    airco2.temperature(20.5)
    airco2.stop()


